### PR TITLE
Add kube_*_annotations metrics

### PR DESCRIPTION
**Pull request by sylr**
**What this PR does / why we need it**:

Adds kube_*_annotations for every existing  kube_*_labels counter parts.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by sylr** [comment]
Hi @brancz, @mrueg 👋 

Could I have a first feedback about this ?

Thank you.

**Comment by brancz** [comment]
I'm generally happy with this. cc @lilic as she lead the label allow work

(I tried to approve the CI run, but github is returning 500 errors consistently every time I press it)

**Comment by lilic** [comment]
I get the same error yes, but works on other PRs. Can you maybe force push @sylr, thanks!

**Comment by sylr** [comment]
I merged master into this branch to trigger the CI.

**Comment by siegenthalerroger** [comment]
I'm just going to leave a link to the relevant issue here so that the PR is correctly linked to the issue with lots of discussion. #941 

Furthermore it's probably important to get @sylr 's thoughts on #1495 as it appears to solve the same use-case but more closely following the design of the label metric

**Comment by sylr** [comment]
@lilic can you approve the workflow please ?

File: docs/cli-arguments.md
Comment (logicalhan) [inline docs/cli-arguments.md:41]: personally, i'd put this below the denylist flag, but this is a bit of a nit.

Comment (logicalhan) [reply docs/cli-arguments.md:41]: also `resource'` has a quote next to it..

Comment (lilic) [reply docs/cli-arguments.md:41]: +1 on this

File: internal/store/utils.go
Comment (lilic) [inline internal/store/utils.go:179]:
```suggestion
// createAnnotationKeysValues takes in passed kubernetes annotations and an allowed list in Kubernetes label format
```

File: internal/store/utils.go
Comment (lilic) [inline internal/store/utils.go:181]: What is the differ4ence between this function and `createLabelKeysValues`? Seems like if we passed a string key of `annotation ` and called `mapToPrometheusLabels ` with that key it would work correctly. Or at least refactor the common things into one function, wdyt?

**Review by lilic — COMMENTED** [review]
A couple of comments, thanks for your patience! We can do majority in a follow up as well if you agree to it or in this tone.

File: internal/store/utils.go
Comment (sylr) [reply internal/store/utils.go:181]: Agreed

**Comment by lilic** [comment]
Sorry about that, but do you mind just rebasing, there was a PR merged, otherwise lgtm from my side 🎉 letting @brancz @tariq1890 or @mrueg have a final look as well. If they don't in a week, will merge. :)

**Comment by sylr** [comment]
@lilic, just merged master into it ;)

**Review by lilic — COMMENTED** [review]
/lgtm
/hold

Thank you! Letting another maintainer have a look, otherwise will merge this next week. 

@sylr in a follow up do you mind adding some docs for this feature, often our users complain this doesn't work out of the box for them. Might be good to find a nice place for this in the main README, in a follow up.

**Comment by sylr** [comment]
I've just pushed a commit which should fix the tests.

**Comment by sylr** [comment]
@lilic what about the following paragraph in the README:

    #### Kubernetes labels and annotations
    
    Since kube-state-metrics v2.0.0 kube_*_labels metrics are not populated any more due to the insane metrics cardinality it generated.
    
    From now on, if you want `kube_*_labels` vectors, you must whitelist the labels for which you want kube-state-metrics
    to generate vectors for. Here's an example of cli arguments which will whitelist the given labels for
    the specified kubernetes objects.
    
    ```
    --metric-labels-allowlist=nodes=[kubernetes.io/arch,node.kubernetes.io/instance-type,topology.kubernetes.io/zone,eks.amazonaws.com/capacityType,    eks.amazonaws.com/nodegroup,company.io/dedicated],statefulsets=[app.kubernetes.io/name,app.kubernetes.io/instance],daemonsets=[app.kubernetes.io/    name,app.kubernetes.io/instance],deployments=[app.kubernetes.io/name,app.kubernetes.io/instance],pods=[app.kubernetes.io/name,app.kubernetes.io/    instance]
    ```
    
    The same mechanism exists for annotations since v2.2.0.
    
    ```
    --metric-annotations-allowlist=namespace=[company.io/owner],statefulsets=[company.io/owner],daemonsets=[company.io/owner],deployments=[company.io/    owner],pods=[company.io/owner]
    ```

**Comment by muryoutaisuu** [comment]
Is this gonna be merged anytime soon as promised or should I invest my time on [kube-annotations-exporter](https://github.com/utilitywarehouse/kube-annotations-exporter)?

File: docs/certificatesigningrequest-metrics.md
Comment (dgrisonnet) [inline docs/certificatesigningrequest-metrics.md:5]: Considering that this is a new metric, we should make it `EXPERIMENTAL` for the time being. This comment also applies to the documentations.

File: internal/store/pod.go
Comment (dgrisonnet) [inline internal/store/pod.go:1132]: nit for consistency, these should be defined in variables.

**Review by dgrisonnet — COMMENTED** [review]
A few comments, but otherwise it looks good to me. Thank you for taking the time to look into this long-standing feature request @sylr!

**Comment by dgrisonnet** [comment]
@muryoutaisuu this feature is in the scope of kube-state-metrics, so it is bound to be added as long as we have enough capacity to look into it so I wouldn't advise you to invest time elsewhere. With that said, you could help us get this feature in kube-state-metrics by reviewing this PR :slightly_smiling_face:

File: internal/store/pod.go
Comment (sylr) [reply internal/store/pod.go:1132]: I used the pattern by of the other functions of the file.

File: docs/certificatesigningrequest-metrics.md
Comment (sylr) [reply docs/certificatesigningrequest-metrics.md]: Done

File: internal/store/certificatesigningrequest.go
Comment (sepich) [inline internal/store/certificatesigningrequest.go:43]: As asked in https://github.com/kubernetes/kube-state-metrics/pull/1495#issuecomment-896964375 I'd suggest to add `MetricFamily` only when `allowAnnotationsList` is provided for this k8s Kind. For example for `csr` it would be: https://github.com/kubernetes/kube-state-metrics/pull/1495/files?file-filters%5B%5D=.go#diff-ee6d72cabf35aed3a2691f1505a707f49a38e752478c98a66d1dfc5ec6b16e51R90
Consider starting ksm with:
`--metric-labels-allowlist=pods=[team] --metric-annotations-allowlist=namespaces=[channel]`
Which would create following metrics for each k8s Kind:
```
kube_<kind>_labels{namespace="ns",<kind>="name"} 1
kube_<kind>_annotations{namespace="ns",<kind>="name"} 1
```
On small testing cluster I'm getting:
```
$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | head
kube_daemonset_annotations{namespace="kube-system",daemonset="local-volume-provisioner"} 1
kube_daemonset_annotations{namespace="kube-system",daemonset="kube-router"} 1
kube_daemonset_annotations{namespace="kube-system",daemonset="node-exporter"} 1
kube_daemonset_annotations{namespace="ceph",daemonset="csi-rbdplugin"} 1
kube_daemonset_annotations{namespace="kube-system",daemonset="filebeat"} 1
kube_daemonset_labels{namespace="ceph",daemonset="csi-rbdplugin"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="filebeat"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="local-volume-provisioner"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="kube-router"} 1
kube_daemonset_labels{namespace="kube-system",daemonset="node-exporter"} 1

$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | wc -l
803
```
These metrics have no sense and only lead to increase of cardinality. From cli args I only interested in `kube_pod_labels` and `kube_namespace_annotations`.
Alternative implementation from https://github.com/kubernetes/kube-state-metrics/pull/1495 would yield:
```
$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | head
kube_namespace_annotations{namespace="vault"} 1
kube_namespace_annotations{namespace="cert-manager"} 1
kube_namespace_annotations{namespace="default",annotation_channel="@UQQ184EHF"} 1
kube_namespace_annotations{namespace="kube-node-lease"} 1
kube_namespace_annotations{namespace="kube-system",annotation_channel="#it-sre-alerts"} 1
kube_namespace_annotations{namespace="it",annotation_channel="#it-sre-alerts"} 1
kube_namespace_annotations{namespace="ceph"} 1
kube_namespace_annotations{namespace="kata-agents",annotation_channel="#it-sre-alerts"} 1
kube_namespace_annotations{namespace="kuberos"} 1
kube_namespace_annotations{namespace="kube-public"} 1

$ curl -s localhost:8080/metrics | egrep '^[^#]+.*(_labels|_annotations)' | wc -l
90  #all pods returns labels
```
Which is more in line with `--metric-allowlist` usage.
In case starting ksm without `--metric-labels-allowlist/--metric-annotations-allowlist` cli arguments set, current implementation would only increase unneeded cardinality. New metrics should be disabled by default https://github.com/kubernetes/kube-state-metrics/issues/941#issuecomment-729560572
> By default this would be all disabled and users could just enable annotations they are interested in.

**Review by sepich — CHANGES_REQUESTED** [review]

File: internal/store/certificatesigningrequest.go
Comment (sylr) [reply internal/store/certificatesigningrequest.go:43]: I hear you but I made this PR so that it replicates the behaviour of `kube_<kind>_labels` metrics.

What you want makes sense to me but I think that it probably should be a separate PR.

@lilic, @dgrisonnet, care to comment about this ?

Comment (dgrisonnet) [reply internal/store/certificatesigningrequest.go:43]: I think it would make sense to hide the `labels` and `annotations` metrics if the allowlists are empty :+1:. Having it changed in a separate PR would make more sense since this effort is already huge and it will only make things harder to review.

File: docs/horizontalpodautoscaler-metrics.md
Comment (dgrisonnet) [inline docs/horizontalpodautoscaler-metrics.md:5]:
```suggestion
| kube_horizontalpodautoscaler_annotations              | Gauge       | `horizontalpodautoscaler`=&lt;hpa-name&gt; <br> `namespace`=&lt;hpa-namespace&gt; | EXPERIMENTAL |
```

File: internal/store/pod.go
Comment (dgrisonnet) [reply internal/store/pod.go:1138]: Still, would you mind changing it so that we have a more consistent codebase? Labels metrics of this same file should also be put in variables.

Comment (sylr) [reply internal/store/pod.go:1138]: I feel this is something to do in another PR. This one is already huge enough.

**Comment by sylr** [comment]
Guys, could we move forward and merge this ?

The size of it PR make it so basically any other PR merged onto master will create conflicts, I already performed two merge/conflicts resolutions.

**Review by dgrisonnet — APPROVED** [review]
/lgtm

**Comment by mrueg** [comment]
/hold cancel
/lgtm
Thanks for your contribution @sylr !

**Pull request merged**
